routing/router.go

```diff
@@ -1620,131 +1665,208 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
+			// startup the ID of the payment that we attempted to send,
+			// such that we can query the Switch for its whereabouts. The
+			// route is needed to handle the result when it eventually
+			// comes back.
+			err = r.cfg.Control.RegisterAttempt(payment.PaymentHash, attempt)
+			if err != nil {
 				return [32]byte{}, nil, err
 			}
 
-			lastError = err
-			continue
+			log.Tracef("Attempting to send payment %x (pid=%v), "+
+				"using route: %v", paymentHash, paymentID,
+				newLogClosure(func() string {
+					return spew.Sdump(route)
+				}),
+			)
+
+			// Send it to the Switch. When this method returns we assume
+			// the Switch successfully has persisted the payment attempt,
+			// such that we can resume waiting for the result after a
+			// restart.
+			err = r.cfg.Payer.SendHTLC(
+				firstHop, paymentID, htlcAdd,
+			)
+			if err != nil {
+				log.Errorf("Failed sending attempt %d for payment "+
+					"%x to switch: %v", paymentID, paymentHash, err)
+
+				// We must inspect the error to know whether it was
+				// critical or not, to decide whether we should
+				// continue trying.
+				finalOutcome := r.processSendError(
+					paySession, route, err,
+				)
+				if finalOutcome {
+					return [32]byte{}, nil, err
+				}
+
+				lastError = err
+
+				// We make another payment attempt.
+				attempt = nil
+				continue
+			}
+
+			log.Debugf("Payment %x (pid=%v) successfully sent to switch",
+				paymentHash, paymentID)
 		}
 
 		// Using the created circuit, initialize the error decrypter so we can
 		// parse+decode any failures incurred by this payment within the
 		// switch.
+		_, circuit, err := generateSphinxPacket(
+			&attempt.Route, paymentHash[:], attempt.SessionKey,
+		)
+
 		errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
 			OnionErrorDecrypter: sphinx.NewOnionErrorDecrypter(circuit),
 		}
 
 		// Now ask the switch to return the result of the payment when
 		// available.
 		resultChan, err := r.cfg.Payer.GetPaymentResult(
-			paymentID, errorDecryptor,
+			attempt.PaymentID, errorDecryptor,
 		)
-		if err != nil {
+		switch {
+
+		// If this payment ID is unknown to the Switch, it means it was
+		// never checkpointed and forwarded by the switch before a
+		// restart. In this case we can safely send a new payment
+		// attempt, and wait for its result to be available.
+		case err == htlcswitch.ErrPaymentIDNotFound:
+			log.Debugf("Payment ID %v for hash %x not found in "+
+				"the Switch, retrying.", attempt.PaymentID,
+				paymentHash)
+
+			// Reset the attempt to indicate we want to make a new
+			// attempt.
+			attempt = nil
+			continue
+
+		// A critical, unexpected error was encountered.
+		case err != nil:
 			log.Errorf("Failed getting result for paymentID %d "+
-				"from switch: %v", paymentID, err)
+				"from switch: %v", attempt.PaymentID, err)
+
 			return [32]byte{}, nil, err
 		}
 
+		// THe switch knows about this payment, we'll wait for a result
```

Comment:
THe -> The

Reply:
Fixed: https://github.com/lightningnetwork/lnd/pull/2761/commits/2cd2f3125678578d85dde4f6e54348d23ff99a2f